components/Input.tsx

```diff
@@ -1,41 +1,126 @@
+  FieldValues,
+  UseFormGetValues,
+  UseFormRegister,
+  UseFormSetError,
+} from "react-hook-form";
+import { postEmailIsValid } from "@/lib/api";
 import eyeOff from "@/public/eye-off.png";
 import eyeOn from "@/public/eye-on.png";
+import { InputType } from "@/types/commonTypes";
 import styles from "./Input.module.css";
 
-function Input() {
-  const [isError, setIsError] = useState(false);
-  const [isEyeOn, setIsEyeOn] = useState(false);
+interface Props {
+  placeholder: string;
+  type: InputType;
+  register: UseFormRegister<FieldValues>;
+  errors: FieldErrors<FieldValues>;
+  setError: UseFormSetError<FieldValues>;
+  getValues: UseFormGetValues<FieldValues>;
+}
+
+const options = {
+  signInEmail: {
+    required: "이메일을 입력해주세요",
+  },
+  signInPassword: {
+    required: "비밀번호를 입력해주세요",
+  },
+  signUpEmail: {
+    required: "이메일을 입력해주세요",
+    validate: {
+      isValidEmail: async (email: FieldValues) => {
+        try {
+          const response = await postEmailIsValid({ email });
+          return response && true;
+        } catch (errorMessage) {
+          return errorMessage;
+        }
+      },
+    },
+  },
+  signUpPassword: {
+    required: "비밀번호를 입력해주세요",
+    minLength: {
+      value: 8,
+      message: "비밀번호는 영문, 숫자 조합 8자 이상 입력해 주세요",
+    },
+  },
+};
+
+const Type = {
+  email: ["signInEmail", "signUpEmail"],
+  password: ["signInPassword", "signUpPassword", "signUpPasswordCheck"],
+};
+
+function Input({ placeholder, type, register, errors, getValues }: Props) {
+  const myOptions = {
+    ...options,
+    signUpPasswordCheck: {
+      validate: {
+        isSamePassword: (passwordCheck: string) => {
+          const values = getValues();
+          const password = values["signUpPassword"];
+          return passwordCheck === password || "비밀번호가 일치하지 않아요";
+        },
+      },
+    },
+  };
+  const { ...rest } = register(type, myOptions[type]);
+  const [isPasswordVisible, SetIsPasswordVisible] = useState(false);
 
-  const inputClassName = isError
+  const inputClassName = errors[type]?.message
     ? `${styles.input} ${styles.error}`
     : styles.input;
 
   const handleEyeClick = () => {
-    setIsEyeOn(!isEyeOn);
+    SetIsPasswordVisible(!isPasswordVisible);
   };
 
-  return (
-    <div className={styles.container}>
-      <div className={styles.innerContainer}>
+  if (Type.email.includes(type))
+    return (
+      <div className={styles.container}>
         <input
           className={inputClassName}
-          type={isEyeOn ? "text" : "password"}
-          placeholder="내용 입력"
-        />
-        <Image
-          className={styles.img}
-          src={isEyeOn ? eyeOn : eyeOff}
-          onClick={handleEyeClick}
-          alt="eye-toggle"
+          type="email"
+          placeholder={placeholder}
+          {...rest}
         />
+        {errors[type] && (
+          <div className={styles.errorMessage}>
+            {errors[type]?.message as string}
+          </div>
+        )}
```

Comment:
- 이부분도 마찬가지로, 타입단언은 위험합니다.
- 이미 조건부터 렌더링하고 있기 때문에, optional chaining하지 않아도 됩니다.
```suggestion
        {errors[type] && (
          <div className={styles.errorMessage}>
            {errors[type].message}
          </div>
        )}
```

Reply:
마찬가지로 리뷰 남기지 않은 부분도 수정해 주세요!